ddtrace/tracer/remote_config.go

```diff
@@ -23,17 +24,93 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
 )
 
+// configData represents one config file received from Remote Config.
 type configData struct {
-	Action        string    `json:"action"`
-	ServiceTarget target    `json:"service_target"`
-	LibConfig     libConfig `json:"lib_config"`
+	Action        string      `json:"action"`
+	ServiceTarget target      `json:"service_target"`
+	K8sTargetV2   k8sTargetV2 `json:"k8s_target_v2"`
+	LibConfig     libConfig   `json:"lib_config"`
 }
 
 type target struct {
 	Service string `json:"service"`
 	Env     string `json:"env"`
 }
 
+type k8sTargetV2 struct {
+	ClusterTargets []clusterTarget `json:"cluster_targets"`
+}
+
+type clusterTarget struct {
+	ClusterName       string   `json:"cluster_name"`
+	Enabled           bool     `json:"enabled"`
+	EnabledNamespaces []string `json:"enabled_namespaces"`
+}
+
+// priority returns the order in which this config should be applied, relative
+// to other configs. The more specific the config's targeting of the current
+// process is, the higher the priority; configs with higher priority override
+// configs with lower priority.
+func (c configData) priority() int {
+	isSingleEnvironment := c.ServiceTarget.Env != "" && c.ServiceTarget.Env != "*"
+	isSingleService := c.ServiceTarget.Service != "" && c.ServiceTarget.Service != "*"
+	isClusterTarget := len(c.K8sTargetV2.ClusterTargets) > 0
+
+	if isSingleEnvironment && isSingleService {
+		return 5
+	}
+
+	if isSingleService {
+		return 4
+	}
+
+	if isSingleEnvironment {
+		return 3
+	}
+
+	if isClusterTarget {
+		return 2
+	}
+
+	return 1
+}
+
+// mergeConfigsByPriority sorts the configs by priority (i.e. targeting
+// specificity) and merges them into a single libConfig. For each field, the
+// value from the highest-priority config with a non-nil value for that field is
+// used.
+func mergeConfigsByPriority(configs []configData) libConfig {
+	if len(configs) == 0 {
+		return libConfig{}
+	}
+
+	sort.Slice(configs, func(i, j int) bool {
+		return configs[i].priority() < configs[j].priority()
+	})
+
+	merged := libConfig{}
+
+	for _, cfg := range configs {
+		if cfg.LibConfig.Enabled != nil {
+			merged.Enabled = cfg.LibConfig.Enabled
+		}
+		if cfg.LibConfig.SamplingRate != nil {
+			merged.SamplingRate = cfg.LibConfig.SamplingRate
+		}
+		if cfg.LibConfig.TraceSamplingRules != nil {
+			merged.TraceSamplingRules = cfg.LibConfig.TraceSamplingRules
+		}
+		if cfg.LibConfig.HeaderTags != nil {
+			merged.HeaderTags = cfg.LibConfig.HeaderTags
+		}
+		if cfg.LibConfig.Tags != nil {
+			merged.Tags = cfg.LibConfig.Tags
+		}
```

Comment:
Do you worry about this rotting when more stuff is added? Could we extract this merging logic into a method and then do something more methodical to test it? I wonder if there's some sort of test we could write that would make it straightforward to detect when new fields are added to libConfig but not to the merge method